File: post-processing-service/scripts/start-service.mjs
import { spawnSync } from 'node:child_process';
import { loadPostProcessingConfig } from '../config/serviceConfig.mjs';
import { createPostProcessingServer } from '../api/server.mjs';
import { ensureFaceModel } from '../setupModel.mjs';

const defaultToken = 'dev-internal-token-change-in-production-32bytes';

function freePort(port) {
  if (process.platform === 'win32') {
    try {
      const result = spawnSync('netstat', ['-ano', '-p', 'tcp'], { encoding: 'utf8', windowsHide: true });
      if (result.status === 0 && result.stdout) {
        const lines = result.stdout.split(/\r?\n/);
        const regex = new RegExp(`TCP\\s+\\S+:${port}\\s+\\S+\\s+LISTENING\\s+(\\d+)`, 'i');
        const pids = new Set();
        for (const line of lines) {
          const match = line.trim().match(regex);
          if (match && match[1] && Number(match[1]) !== process.pid) {
            pids.add(match[1]);
          }
        }
        for (const pid of pids) {
          console.log(`Port ${port} is in use by process PID ${pid}. Terminating process...`);
          spawnSync('taskkill', ['/PID', pid, '/T', '/F'], { stdio: 'ignore', windowsHide: true });
        }
      }
    } catch {
      /* ignore cleanup errors */
    }
  }
}

async function main() {
  process.env.POST_PROCESSING_PILOT_ENABLED = process.env.POST_PROCESSING_PILOT_ENABLED ?? 'true';
  process.env.POST_PROCESSING_INTERNAL_TOKEN = process.env.POST_PROCESSING_INTERNAL_TOKEN || defaultToken;

  const config = loadPostProcessingConfig();

  // Free port 6501 if previously occupied
  freePort(config.runtime.port);

  console.log('Preparing Face Landmarker model...');
  try {
    await ensureFaceModel({ config });
    console.log('Model artifact verified.');
  } catch (error) {
    console.warn('Face Landmarker model warning: ' + error.message);
    console.warn('Starting service with model capabilities marked unavailable.');
  }

  const server = await createPostProcessingServer({ config });
  const host = config.runtime.host;
  const port = config.runtime.port;
  const token = config.runtime.internalToken;

  server.on('error', error => {
    if (error.code === 'EADDRINUSE') {
      console.error(`\nError: Port ${port} is already in use by another process.`);
      console.error('Please close any process occupying port ' + port + ' and try again.\n');
    } else {
      console.error('Server error:', error);
    }
    process.exit(1);
  });

  server.listen(port, host, () => {
    console.log('\n=============================================================');
    console.log('   Post-Processing Microservice (Isolated Standalone Mode)');
    console.log('=============================================================');
    console.log(`   URL:          http://${host}:${port}`);
    console.log(`   InternalToken:${token}`);
    console.log(`   Pilot Mode:   ${config.runtime.pilotEnabled ? 'ENABLED' : 'DISABLED'}`);
    console.log('-------------------------------------------------------------');
    console.log('   Available Endpoints (JSON Format):');
    console.log(`   - GET  http://${host}:${port}/health`);
    console.log(`   - GET  http://${host}:${port}/v1/health`);
    console.log(`   - GET  http://${host}:${port}/v1/capabilities`);
    console.log(`   - GET  http://${host}:${port}/v1/metrics`);
    console.log(`   - POST http://${host}:${port}/v1/faceless-previs`);
    console.log(`   - POST http://${host}:${port}/v1/face-landmarks`);
    console.log('=============================================================\n');
    console.log('Ready for isolated cURL / Postman testing. Press Ctrl+C to stop.');
  });

  const shutdown = async () => {
    console.log('\nStopping Post-Processing Service...');
    server.close(() => {
      console.log('Service stopped.');
      process.exit(0);
    });
  };

  process.on('SIGINT', shutdown);
  process.on('SIGTERM', shutdown);
}

main().catch(error => {
  console.error('Failed to start Post-Processing Service:', error);
  process.exit(1);
});

File: post-processing-service/scripts/start_service.py
import os
import re
import sys
import subprocess
from pathlib import Path

SERVICE_ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(SERVICE_ROOT))

from config.service_config import load_post_processing_config, D_APPLICATIONS_ROOT

DEFAULT_TOKEN = "dev-internal-token-change-in-production-32bytes"

def free_port(port: int):
    if sys.platform == "win32":
        try:
            result = subprocess.run(
                ["netstat", "-ano", "-p", "tcp"],
                capture_output=True, text=True, check=False
            )
            if result.returncode == 0 and result.stdout:
                lines = result.stdout.splitlines()
                regex = re.compile(rf"TCP\s+\S+:{port}\s+\S+\s+LISTENING\s+(\d+)", re.IGNORECASE)
                pids = set()
                current_pid = os.getpid()
                for line in lines:
                    match = regex.search(line.strip())
                    if match and match.group(1) and int(match.group(1)) != current_pid:
                        pids.add(match.group(1))
                for pid in pids:
                    print(f"Port {port} is in use by process PID {pid}. Terminating process...")
                    subprocess.run(["taskkill", "/PID", pid, "/T", "/F"], capture_output=True, check=False)
        except Exception:
            pass

def main():
    os.environ["POST_PROCESSING_PILOT_ENABLED"] = os.environ.get("POST_PROCESSING_PILOT_ENABLED", "true")
    os.environ["POST_PROCESSING_INTERNAL_TOKEN"] = os.environ.get("POST_PROCESSING_INTERNAL_TOKEN", DEFAULT_TOKEN)

    config = load_post_processing_config()
    free_port(config.runtime.port)

    print("Preparing Face Landmarker model under D:\\applications...")
    try:
        from scripts.setup_model import ensure_face_model
        ensure_face_model(config)
    except Exception as e:
        print(f"Model initialization warning: {e}", file=sys.stderr)
        print("Starting service with model capabilities marked unavailable.")

    host = config.runtime.host
    port = config.runtime.port
    token = config.runtime.internalToken

    print("\n=============================================================")
    print("   Momelo Post-Processing Service (Python FastAPI Mode)")
    print("=============================================================")
    print(f"   URL:          http://{host}:{port}")
    print(f"   Swagger Docs: http://{host}:{port}/docs")
    print(f"   InternalToken:{token}")
    print(f"   Pilot Mode:   {'ENABLED' if config.runtime.pilotEnabled else 'DISABLED'}")
    print("-------------------------------------------------------------")
    print("   Available Endpoints (100% JSON Format):")
    print(f"   - GET  http://{host}:{port}/health")
    print(f"   - GET  http://{host}:{port}/v1/health")
    print(f"   - GET  http://{host}:{port}/v1/capabilities")
    print(f"   - GET  http://{host}:{port}/v1/metrics")
    print(f"   - POST http://{host}:{port}/v1/faceless-previs")
    print(f"   - POST http://{host}:{port}/v1/face-landmarks")
    print("=============================================================\n")
    print("Ready for isolated cURL / Postman / Swagger testing. Press Ctrl+C to stop.")

    import uvicorn
    uvicorn.run("main:app", host=host, port=port, log_level="info", app_dir=str(SERVICE_ROOT))

if __name__ == "__main__":
    main()
